Add best practice guide (#1919)
aws/eks-anywhere: DEVELOPMENT_BEST_PRACTICE.md

> "Debugging is twice as hard as writing the code in the first place. Therefore, if you write the code as cleverly as possible, you are, by definition, not smart enough to debug it." - _Brian W. Kernighan_

- [SOLID](https://www.digitalocean.com/community/conceptual_articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#single-responsibility-principle) - The Single-responsibility Principle, Open-closed Principle, Liskov Substitution Principle, Interface Segregation Principle and Depdendency Inversion Principle collectively define a guide for writing maintainable code. While often referenced in the context of an object-oriented language, they are applicable to all languages.
- [LoD](https://en.wikipedia.org/wiki/Law_of_Demeter) - The Law of Demeter tells us a construct should talk to their direct dependencies and _only_ their direct dependencies. Reaching to transitive dependencies creates complex layers of interaction that drive toward [spaghetti code](https://en.wikipedia.org/wiki/Spaghetti_code).
- [KISS](https://people.apache.org/~fhanik/kiss.html) - "Keep it simple, stupid" was coined by the US Navy. Systems should be designed as simply as possible.

> - [YAGNI](https://martinfowler.com/bliki/Yagni.html) - "You aren't gonna need it" if you don't have a concrete use-case, so don't write it.

Author: Not sure I agree with this one. Reading it with the suggestion seems odd.

> - [YAGNI](https://martinfowler.com/bliki/Yagni.html) - "You aren't gonna need it" if you don't have a concrete use-case, so don't write it.

Author: Could drop the 'so' entirely?

- [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) - "Don't repeat yourself" suggests you should preference code re-use over duplication. However, you should [avoid hasty abstractions](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction) as the wrong abstraction can be extremely costly to correct. Instead, lean into small amounts of duplication to help identify the right abstractions through multiple use cases and define the pathway to DRY design.

## "clean code that is optimized for change"

What constitutes clean code? We want simple and maintainable, but how do we define that? The following best practices offer niche advice. Combined, they represent our definition of clean code.

### State

> Shared mutable state is believed by many to be the “root of all evil”, or at least the cause of most of the accidental complexity in our code. And “Complexity is the root cause of the vast majority of problems with software today.” - [Mauro Bieg](https://mb21.github.io/blog/2021/01/23/pure-functional-programming-and-shared-mutable-state.html#the-root-of-all-evil)

> Package level state is rarely required. If it is, it should be composed of re-usable constructs defined and exposed from the package itself.

Author: And unexported with the 1 exception of not being able to specify a struct as a const

In avoiding package level state we avoid [hidden dependencies and unintended side effects](https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_package_level_state) stemming from global state mutation. Model loosely coupled components by [declaring your dependencies](#dependencies).

### Names

##### Packages

> Package names help form the first impression of what a package provides and is used to reference the public API of the package. A good package name brings about clarity and purpose for consumers. Package names should be concise and represent the behavior the package provides not the types it contains. For example, `meals` provides `dr

Author: Yes

Avoid stutter in struct and interface names by treating the package name as a namespace and considering how the types are consumed. For example, `drink.DrinkDecorator` should be `drink.Dectorator`.

If you find yourself using an appropriate package name that's commonly used as a variable, consider pluralising the package name.

> Avoid (like the plague) `types`, `interfaces`, `common`, `util` or `base` packages. They don't represent anything cohesive and tend to collect unrelated types and algorithms that often require teething apart as a project progresses due to cyclic dependencies. If a utility style package is required, make it specific. For example a `/utils/cmdutil` package may provide command utility functions.

Author: 🤦🏻 thanks

##### Function & method names

Functions and methods should adequetly describe their behavior. Generally, they should follow a verb-noun form.

##### Variable names

Variable names should be concise and descriptive. Prefer single word names. The further away from the site of declaration a variable is used, the more descriptive it needs to be. For example, a variable named `n` used 30 lines after declaration makes it unnecessarily difficult to reason what it represents at the site of use.

### Abstractions

- Functions & Methods
- Constructs
- Interfaces

> ### <a nane="dependencies"></a> Depencencies

Reviewer: ```suggestion
### <a name="dependencies"></a> Dependencies
```

- Accept interfaces, return structs
- Declare your dependencies

### Package API

> - this might be wrapped up in naming and testing. Do we have more explicit points?

Reviewer: You could always just link https://go.dev/blog/package-names I guess. :D

### Testing

> - test using the `_test` idiom

Reviewer: A reference to https://pkg.go.dev/testing might be helpful.

- Use test to help identify if the package is easy to use for consumers

> - isolating concerns like io

Reviewer: ...and time! Time is a big one in tests in my experience.

### Interfaces

- Define the behavior a _consumer_ expects
- Premature interfaces
- Accept interfaces
- Should be cohesive
- Should be small
- Preferably single method

### Errors

- Describe the action that failed
- Use lower case
- Don't prepend error
- Don't overshare information 

> - Add context if you can

Reviewer: - wrap an underlying error when appropriate (e.g. when they provide more information)

### Boolean logic

- Toggling behavior using boolean values is generally bad
- Create dynamic constructs that can be built without the behavior

### Concurrency

- Prefer synchonous APIs
- Leave it to the caller
- Mask from the caller

### Disambiguating context.Context

- Represents the context of execution
    - Useful for tear down
- There should be 1 and only 1 context
- Should be the first parameter on the primary call path

### Loggers

> - They're just another dependency, inject them.

Reviewer: Just an aside I guess, but I see a lot of package-level loggers in k8s code. I sure don't care for that, so I'm glad to see this.

Author: Yea I don't personally like the `klog` thing (and some other bits) Kubernetes does. Its the easy way out of managing a significant dependency (the logger). I've worked on projects, and you see it frequently in Go projects, where you let the code grow organically and decide to extract some code to a re-usable repository only to find there are package level dependencies 🤦🏻 .

I'll write as much here as an elaboration to 'why'.

###  Comments

- All public APis should be documented
- May seem redundant at times, consider documentation from a `godoc` point of view.
- Packages _should_ contain package level documentation
- What makes a good comment?

### Types

- I forgot what this was for...

### Variable initialization

> - What forms to preference under what circumstances

Reviewer: The guideline I remember reading was to use `var x int` when you aren't sure if you'll ever assign something to the value. Use `x := 0` when you _know_ you will. Applies more to slices and structs than ints I think, but int was shorter/easier to type.

### Channels

- If your API leverages them, inject them
- Document channel behavior

### Returns

- Return early
- Naked returns should be avoided

> ### Panicing

Reviewer: ```suggestion
### Panicking
```

### Names

##### Packages

Package names help form the first impression of what a package provides and is used to reference the public API of the package. A good package name brings about clarity and purpose for consumers. Package names should be concise and represent the behavior the package provides not the types it contains. 

Avoid stutter in struct and interface names by treating the package name as a namespace and considering how the types are consumed. For example, `drink.DrinkDecorator` should be `drink.Decorator`. You may find you have a type named after the package, soemtimes it's appropriate.

If you find yourself using an appropriate package name that's commonly used as a variable, consider pluralising the package name.

Avoid (like the plague) `types`, `interfaces`, `common`, `util` or `base` packages. They don't represent anything cohesive and tend to collect unrelated types and algorithms that often require teething apart as a project progresses due to cyclic dependencies. If a utility style package is required, make it specific. For example a `/utils/cmdutil` package may provide command utility functions.

##### Function & method

> Functions and methods (functions with a receiver parameter) should adequetly describe their behavior. Generally, they should follow a verb-noun form. The types of parameters should help inform the function name.

Author: Parameter types help inform on the functions scope/purpose, not necessarily its name.

```go
func FastenPlanks(tools Toolbox, surface Surface, planks Planks) error {
    var plank Plank
    for ; planks.Count() > 0; plank = planks.Next() {
        if err := tools.Hammer(surface, ); err != nil {
            return err
        }
    }
}
```

Avoid constructing dependencies in-line. If you do construct a dependency in-line because it compliments default usage, provide a way to override it such as with a setter.

#### Loggers

Loggers are sometimes treated as a special depency by making entire projects depend on a package level logger. This _masks_ a crucial dependency that is hard to change without impacting the rest of the code base. To understand why this is bad, consider extracting a package into its own repository  for reusability. Depending on a package level logger, or any other package level dependencies, makes the extraction process unnecessarily difficult as you first need to pull the hidden dependency apart.

> Logger's are not special dependencies, inject them like any other dependency. If you want to minimize the surface area of the default initializing constructor (`New...` funcs) you may construct noop loggers to satisfy the logger dependency and provide a means to override the logger with a setter.

Reviewer 2: "inject" makes me think about dependency injection. Youre not suggested that here are you? might think about a synonym?

Author: That's exactly what I'm suggesting :).

Author: See https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_package_level_state and

Reviewer 2: Ah yes. I'm very much on the same page here. I believe my previous definition of dependency injection was incorrect.

"return structs" is less obvious. It, generally, speaks to initializing constructors where authors should return the struct being constructed as opposed to interface it satisifes. This is largely due to duck typing (["If it walks like a duck and it quacks like a duck, then it must be a duck"](https://en.wikipedia.org/wiki/Duck_typing)).

> Interfaces in Go provide a way to specify the behavior of an object: if something can do this, then it can be used here. - [Effective Go](https://go.dev/doc/effective_go#interfaces_and_types)

Factory functions are not to be confused with intializing constructors. Factory functions typically return abstractions (an interface) satisfied by 1 of several implementations. A factory could be as simple as:

```go
func NewStringer(key string) (fmt.Stringer, error) {
    switch key {
    case "foo":
        return Foo{}, nil
    case "bar":
        return Bar{}, nil
    default:
        return nil, errors.New("unknown key")
    }
}

> type Foo struct {}

Author: Replace with a note on the func for example brevity.

- [LoD](https://en.wikipedia.org/wiki/Law_of_Demeter) - The Law of Demeter tells us a construct should talk to their direct dependencies and _only_ their direct dependencies (don't talk to strangers). Reaching to transitive dependencies creates complex layers of interaction that drive toward [spaghetti code](https://en.wikipedia.org/wiki/Spaghetti_code).
- [KISS](https://people.apache.org/~fhanik/kiss.html) - "Keep it simple, stupid" was coined by the US Navy. Systems should be designed as simply as possible.
- [YAGNI](https://martinfowler.com/bliki/Yagni.html) - "You aren't gonna need it" if you don't have a concrete use-case, don't write it.
- [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) - "Don't repeat yourself" suggests you should preference code re-use over duplication. However, you should [avoid hasty abstractions](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction) as the wrong abstraction can be extremely costly to correct. Instead, lean into small amounts of duplication to help identify the right abstractions through multiple use cases and define the pathway to DRY design.

> "Leave code better than you found it"

> Generally, you want to fix problems as they're identified. This includes micro refactors to accommodate new code that causes existing abstractions to, possibly, no longer make sense. Resist bolting on new code that is uncomplimentary to existing code.

Reviewer: This may warrant a comment on how it's great to make this change, but probably in a different PR. Like code, PRs should aim to follow the SRP. :)

Author: Yea good callout. Basically commentary on _how_ to do this.

Reviewer 3: My advice on this is usually: if the refactor idea can stand on itself, create a separate PR with just that.

But how to decide if the idea stands on its own? A somehow ok rule of thumb: if you would write a separate commit for the refactor, then it might be worth its own PR.

Author: Addressed.

> Finally, we have delivery commitments. Larger more intentional refactors may be required to fix a problem you've identified and attempting to tackle it may delay your deliverable. In such a case, write down your ideas and start a conversation with the wider team. 

Reviewer 3: I would vote against mentioning "creating tickets". Compromising on quality and postponing refactors is one answer to the deliverable problem, but in my opinion, it shouldn't be the default one. I'm not saying that's what you were proposing, but I think mentioning in this doc can give the wrong idea.

I think the key part of this paragraph is "start a conversation with the wider team" (maybe we could expand on it?). We should try to always talk when this kind of thing happen, both to be able to find the right compromise (usually quality, time or scope) for each situation (one the team feels comfortable with) and to be able to detect underlying issues.

Defaulting to just postponing the work can hide other issues:
* We are underestimating tasks. This is quite common and perfectly normal, I'm confident we all do it, at least I definitely do it. But the only way to get better at it is to recognize when it happens and learn from it. Plus, I think we have to get used to accommodate some "rework" in our estimations. The codebase has grown to a point where almost nothing is greenfield anymore, so improving it as we go should be an intrinsic part of our day to day work.
* We produce "unclean" code. When that happens, it's usually difficult to maintain and expand, hence requiring significant refactors on every new feature. I personally don't think this has been a major problem for us but we should stay on top of it.
* We are accumulating too much tech debt. This has similar symptoms as the previous one but it's caused by little inoffensive changes accumulated over time.

Moreover, always postponing the work is unsustainable. The more we do it, the more tech debt we accumulate. And the more tech debt we accumulate, the more refactor will be needed to add new features, which feeds back into the same loop. This feedback loops will continue to the point where we start spending more time on refactors than on improving the product.

> Finally, we have delivery commitments. Larger more intentional refactors may be required to fix a problem you've identified and attempting to tackle it may delay your deliverable. In such a case, write down your ideas and start a conversation with the wider team. 

Reviewer 4: yeah no I get you, I'm mostly just saying that when we 'write it down somewhere' as mentioned explicitly above that place should often be a github issue

```go
// Foo and Bar satisfy fmt.Stringer
func NewStringer(key string) (fmt.Stringer, error) {
    switch key {
    case "foo":
        return Foo{}, nil
    case "bar":
        return Bar{}, nil
    default:
        return nil, errors.New("unknown key")
    }
}
```

Factories generally shouldn't contain complex construction logic. That's the job of the initializing constructor for the concrete type.

##### Loggers

Loggers are sometimes treated as a special depency by making entire projects depend on a package level logger. This _masks_ a crucial dependency that is hard to change without impacting the rest of the code base. To understand why this is bad, consider extracting a package into its own repository for reusability. Depending on a package level logger, or any other package level dependencies, makes the extraction process unnecessarily difficult.

Logger's are not special dependencies, inject them like any other dependency. If you want to minimize the surface area of the default initializing constructor (`New...` funcs) you may construct noop loggers to satisfy the logger dependency and provide a means to override the logger with a setter.

> ### Testing

Reviewer 4: could be worth noting when to implement functional/end-to-end tests in this section

Author: I'd be keen to call out a separate section for testing types. This topic can be quite large and often involves turmoil due to a lack of definition around what constitutes an integration test vs unit vs e2e etc.

Author: Addressed.

- laptops with less than 16" wide monitors are in abundance.
- side-by-side comparison of files is useful.
- horrizontal scrolling is annoying.
- some developers _need_ larger font sizes.

When developing code, we ask you to take these points into consideration and not to create obnoxiously long lines. Compliment exisiting code. Configure your IDE to plot margins so it may aid your decision. A good rule of thumb is the 100 characters mark give or take 20.

### File naming

All `go` file names must follow `snake_case`. Using snake case compliments the Go compiler that expects test files to be appended with `_test`.

## References

These references are the basis for this document.

- Practical Go - Dave Cheney - https://dave.cheney.net/practical-go/presentations/gophercon-israel.html
- Best Practices - Peter Bourgon - https://peter.bourgon.org/go-best-practices-2016/
- Effective Go - The Go Team - https://go.dev/doc/effective_go
- Code Review Comments - The Go Community - https://github.com/golang/go/wiki/CodeReviewComments

> ## To do

Reviewer 4: would love to see more about logging -- when, what, writing concise/clear log messages; really important for building for operations, especially since we'll be debugging a lot of this stuff 'blind' without access to the cluster in most cases

Author: Yea great point. I've written a 1 pager on logging previously because of how important it is. I'll add another section in the to do.

Author: Added the todo.

Package level state is rarely required. If it is, it should be composed of re-usable constructs defined and exposed from the package itself.

In avoiding package level state we avoid [hidden dependencies and unintended side effects](https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_package_level_state) stemming from global state mutation. Model loosely-coupled components by [declaring your dependencies](#dependencies).

### Names

##### Packages

Package names help form the first impression of what a package provides and is used to reference the public API of the package. A good package name brings about clarity and purpose for consumers. Package names should be concise and represent the behavior the package provides not the types it contains. 

Avoid stutter in struct and interface names by treating the package name as a namespace and considering how the types are consumed. For example, `drink.DrinkDecorator` should be `drink.Decorator`. You may find you have a type named after the package, soemtimes it's appropriate.

> If you find yourself using an appropriate package name that's commonly used as a variable, consider pluralising the package name.

Author: Sure. I'll add examples from the stdlib also.

> If you find yourself using an appropriate package name that's commonly used as a variable, consider pluralising the package name.

Author: Addressed.

Avoid [relentlessly] `types`, `interfaces`, `common`, `util` or `base` packages. These names offer no indication on what the package provides beyond 'a bundle of stuff'. With time, these types of package turn into dumping grounds containing unrelated logic that can be the root cause for cyclic dependencies.  If a utility style package is required, make it specific. For example a `/utils/cmdutil` package may provide command utility functions.

##### Function & method

Functions and methods (functions with a receiver parameter) should adequetly describe their behavior. Generally, they should follow a verb-noun form. The types of parameters should help inform the function name.

##### Variable

> Variable names should be concise and descriptive. Prefer single word names. The further away from the site of declaration a variable is used, the more descriptive it needs to be. For example, a variable named `n` used 30 lines after declaration makes it unnecessarily difficult to reason what it represents at the site of use.

Reviewer 5: Is this doc serving for general Go practice or EKS Anywhere best practice?

If latter, can we expand a bit for recommended eks-a naming convention? e.g. long vs short:`kubeadmControlPlane` vs `kcp`; capitalize acronym or not: `capi` vs `CAPI`, `ip` vs `IP`, etc.

Author: Its both but its scope only extends to EKS-A :). 

Re naming, we can expand the variable naming section to elaborate on recommendations. We should probably end with 'if in doubt, prefer clarity over brevity'.

Author: Addressed.

### <a name="dependencies"></a> Dependencies

> Functions should explicitly declare their dependencies as parameters. In doing so, they honor the [dependency inversion principle](https://www.digitalocean.com/community/conceptual_articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#dependency-inversion-principle) and provide the necessary flexibility to write loosely coupled programs.

Reviewer 3: This is totally correct but, reading it, gives me the impression that defining dependencies as arguments is enough to follow dependency inversion (even though you are not saying that explicitly).

Maybe this is just me (in that case ignore me), but I would maybe specify that doing this is necessary when following Dep Inversion but it's not enough.

Author: Its intentionally worded as 'honor'ing not satisfying for the reasons you purport. Perhaps there's a better word here that will clarify it, or perhaps the back half of the sentence is misleading.

Reviewer 3: Totally, as I said, your wording is completely accurate. But I just wonder if I know this because I already knew the difference between one and the other.

One option is to add something like:

```suggestion
Functions should explicitly declare their dependencies as parameters. In doing so, they honor (but don't completely satisfy) the [dependency inversion principle](https://www.digitalocean.com/community/conceptual_articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#dependency-inversion-principle) and provide the necessary flexibility to write loosely coupled programs.
```

Maybe this is just me, if no one else comments about this, feel free to ignore me.

```go
func FastenPlanks(tools Toolbox, surface Surface, planks Planks) error {
    for ; planks.Count() > 0;  {
        if err := tools.Hammer(surface, planks.Next()); err != nil {
            return err
        }
    }
}
```

Avoid constructing dependencies in-line. If you do construct a dependency in-line because it compliments the most common usage, provide a way to override it such as with a setter.

##### "Accept interfaces, return structs"

The Go community talks of the "accept interfaces, return structs" addage. The meaning of this idiom can be hard to pin down, herein lies some clarity.

> "Accept interfaces" speaks to the Dependency Inversion Principle (D in 'SOLID'). When a function declares dependent behavior as an interface parameter it gives control of _how_ that behavior is satisfied to the consuming logic. This is the foundation of flexible code.

Reviewer 3: Repeating myself here a bit but: using interfaces in go is key for the Dependency Inversion Principle, but not enough. What that interface represents and the signature of its methods play a key role as well.

Maybe this is something we can elaborate more on the interfaces section: using an interface is not enough to achieve loosely coupling.

Author: I'm not sure I agree. Dependency inversion can be achieved with an interface parameter. I don't think we should complicate it beyond that. Whether your _design_ is providing increased flexibility is related but different. Your interface designs speaks to the _degree_ of flexibility but the degree of flexibility goes beyond dependency inversion.

Reviewer 3: My bad, I think I didn't explain myself correctly. I might do better with an example:

```go
# mysql/auth.go

type Credentials struct {
   User, Password string
}

func (a *Authenticator) Authenticate(ctx context.Context, creds Credentials) error {...}

# user/registry.go

type DBConnector interface {
    Authenticate(ctx context.Context, creds Credentials) error
}

type Registry struct {
   dBConnector DBConnector
}

func NewRegistry(dBConnector DBConnector) *Registry {...}

func (r *Registry) Start(ctx context.Context) error {
   ...
   err := r.Authenticate(ctx, mysql.Credentials{User: user, Password: password})
   ...
}
```

Here, the registry is taking its dependency as an interface, however, it doesn't achieve dependency inversion because it still depends on a type from the concrete implementation. This is what I meant that just using an interface is not enough. You need to design the interface appropriately, for example:

```go
# mysql/auth.go
func (a *Authenticator) Authenticate(ctx context.Context, user, password string) error {...}

# user/registry.go
type DBConnector interface {
    Authenticate(ctx context.Context, user, password string) error
}

func (r *Registry) Start(ctx context.Context) error {
   ...
   err := r.Authenticate(ctx, user, password)
   ...
}
```

Author: I think that example alludes to different degrees of coupling which, while related, isn't strictly required to satisfy the dependency inversion principle. At the most basic level, dependency inversion can still be achieved for the purpose of testability by accepting an interface parameter. That interface may provide behavior that accepts types associated with a specific package. Yes, its advantageous to only use standard library types on interfaces and when you model complete abstractions a set of shared types but that again starts to delve into software design.

Given the contention I think I'll alter the wording by replacing any reference to DIP with 'loose coupling' or similar. The DIP commentary isn't necessary to convey the point.

Reviewer 3: Sounds good, I'm ok with removing the DIP reference

Author: Addressed.

Factory functions are not to be confused with intializing constructors. Factory functions typically return abstractions (an interface) satisfied by 1 of several implementations. A factory could be as simple as:

```go
// Foo and Bar satisfy fmt.Stringer
func NewStringer(key string) (fmt.Stringer, error) {
    switch key {
    case "foo":
        return Foo{}, nil
    case "bar":
        return Bar{}, nil
    default:
        return nil, errors.New("unknown key")
    }
}
```

Factories generally shouldn't contain complex construction logic. That's the job of the initializing constructor for the concrete type.

##### Loggers

Loggers are sometimes treated as a special depency by making entire projects depend on a package level logger. This _masks_ a crucial dependency that is hard to change without impacting the rest of the code base. To understand why this is bad, consider extracting a package into its own repository for reusability. Depending on a package level logger, or any other package level dependencies, makes the extraction process unnecessarily difficult.

> Logger's are not special dependencies, inject them like any other dependency. If you want to minimize the surface area of the default initializing constructor (`New...` funcs) you may construct noop loggers to satisfy the logger dependency and provide a means to override the logger with a setter.

Reviewer 3: I'm not against this at all. But due to the fact that all the cli code violates this advice, if we want to push for it we should probably pair it with a proposal on how to implement this on new code (eg. inject from the dependency factory to the constructors...) and a plan to little by little refactor existing code.

We should definitely do this separately to this PR but maybe worth adding a note here.

I get this doc aims to help us write better code in general and not only for the CLI. But since that's the piece of software we spent most time with, I figured it might be a good idea to bridge that gap.

Author: Definitely. I think the CLI has lots of flaws, actually, that need tidying up - this is 1 of those flaws.

Author: I'm making notes on what we need to follow up on and have included this. I'll create a well written issue describing the problem and referencing this guide once its merged.

Author: Addressed.

### Testing

Software development is costly and hazardous. Tests provide example usage, guard rails, and assurances. Without tests, you cannot have high quality code. For this reason, all code we spend time designing and writing is worth testing. Tests should:

- focus on the public API of a packge.
- use the `_test` idiom for testing files.
- treat the package as a black box.

> - cover at least 80% of the public API.

Reviewer 3: clarification: testing 80% of the public functions/methods (eg. just calling those methods) or covering 80% of the code accessible through public APIs?

Author: Good point. 80% line coverage. Your code should be flexible enough to enable that. I've never had a problem and even strived for higher, 90%. With respect to the public API, we probably want something like 90%+ covered.

I think its worth calling out I chose this number based on personal experience. We don't have to stick to it but I'm definitely advocating for it.

Reviewer 3: If this gets approved (which I do), let's talk about adding a rule to codecov.

An example of code designed to be dynamically constructed may include the [decorator](https://refactoring.guru/design-patterns/decorator) pattern. It may use a [builder](https://refactoring.guru/design-patterns/builder) or an [abstract factory](https://refactoring.guru/design-patterns/abstract-factory) with a builder to construct the logic using boolean flag configuration read from CLI flags. There are many more combinations and, generally, you want to consider your problem space and design a piece of code that makes sense for your problem first.

### Concurrency

Concurrency introduces harder to rationalize execution relative to serialized programs and is easy to get wrong. Concurrency doesn't necessarily speed up a program either. With these points in mind, prefer synchronous APIs instead leaving concurrency to the caller. Not only does this eliminate a whole host of problems, it ensures your APIs are predictable and easier to use.

Dave Cheney has made some excellent [comments](https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#concurrency) on concurrency in Go.

#### Channels

> If you find the need for channels prefer accepting them as parameters. Accepting channels as a parameter allows consumers to specify channel properties, such as buffered or unbuffered, creating a more flexible API. 

Reviewer 3: Clarification: channels for input, channels for input or both?
I would agree with channels for input, I would disagree with channels for output. At least as general advice.

Author: Both. Even more so for output. As the doc notes, allowing consumers to pass the channel lets them control buffering which, as a consumer, is hugely important if you have an API experiencing bursts while you try to provide a more consistent throughput from your API. 

We should also add a note about conditionality of the channel.

Reviewer 3: If you feel strongly about this, go ahead, I'm ok with it 🙂 

However, I don't necessarily agree this should always be the answer. What you are proposing does in fact offer a more flexible API but in some cases the price to pay would be a more complicated implementation.

To you example, burst smoothing can be easily implemented (kind of as a decorator) even if the concurrent code method returns a non buffered channel. I'm not saying it has to be that way, just that imo this is more a tradeoff topic than a clear rule of thumb.

Clearly document channel ownership thereby documenting who is responsible for closing it. If there are multiple writers, no-one can close it therefore no-one owns it. Note channels don't _need_ to be closed, they will be garbage collected once the necessary scoping conditions are met.

Clearly document channel behavior. For example, what does your algorithm do if it can't send data on a channel? Will it drop data, block until the channel is writable, or something else?

Avoid using channels for data mutation operations where a traditional mutex would suffice. 

###  Comments

> Functions and variables should be named so that their purpose is clear. However, good naming and well structured code isn't justification to not document APIs. All public package APIs should be documented including those that may seem trivial. This, at minimum, helps with code documentation generation. Additionally, each package should have a clear purpose that can be documented with examples using the [package documentation feature of Go](https://go.dev/doc/effective_go#commentary).

Reviewer 3: Note: If everyone agrees with this, we should probably look into adding some linter to prevent new public entities/functions from being added without doc comments.

Author: Made a note.

> Through good documentation that we reflect on and update regulalry, we can clearly communicate with maintainers what the original intent is making decisions easier when new capabilities are required.

Reviewer 3: ```suggestion
Through good documentation that we reflect on and update regularly, we can clearly communicate with maintainers what the original intent is making decisions easier when new capabilities are required.
```

### Returns

Return early instead of nesting deeply. The `else` statement is rarely required when following this rule. For example:

```go
func foo(bar string) error {
    if bar == "" {
        return errors.New("bar is empty")
    }

    // do something

    return nil
}
```

In returning early, we also create clearer code that has lower [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) that can be the cause for bugs.

Never use naked returns. Go allows return types to be specified with an identifier. The identifier is a scoped variable in the context of the function and its value can be set during program execution. Doing so, however, often creates unnecessraily difficult to rationalize logic.

> Avoid named returns. If you don't specify names for your return values you can't accidentally use naked returns. 

Reviewer 3: I disagree with this. I think named returns are very useful when returning things like `(string, string)` and I think we should encourage them.

I would agree with avoiding them when unnecessary (eg. `ReadFileName() (string, error)`). Also 100% agree with "Never use naked returns".

Author: Thats why this is `avoid`. Common sense needs to be used and if you have a function that returns the same type twice then clearly named returns help clarify whats what. Thats pretty much the only case I thought about that warrants named returns and caused my to choose the wording 'avoid'.

Reviewer 3: What would you think about replacing "avoid" with something like:
```suggestion
Use named returns only if strictly necessary for disambiguation (for example, a function that returns the same type twice). If you don't specify names for your return values you can't accidentally use naked returns. 
```
